include/zephyr/sys/kref.h

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2023 Michael Zimmermann
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_KREF_H_
+#define ZEPHYR_INCLUDE_SYS_KREF_H_
+
+#include <stdbool.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/toolchain.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup kref_apis Reference count APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/**
+ * @brief An atomic reference count.
+ *
+ * The main use case is to share the same data between several users and free
+ * it when the count drops to 0. This can also be used on SMP systems.
+ */
+struct k_ref {
+	atomic_t refs;
+};
+
+/** @brief Initialize k_ref at compile-time. */
+#define K_REF_INIT()                                                                               \
+	{                                                                                          \
+		.refs = ATOMIC_INIT(1),                                                            \
+	}
+
+/** @brief Initialize k_ref at runtime. */
+void k_ref_init(struct k_ref *ref);
+
+/**
+ * @brief Increase the reference count.
+ *
+ * @return `ref` is successful or
+ *         NULL if the reference count was 0 already.
+ */
+struct k_ref *k_ref_get(struct k_ref *ref);
```

Comment:
Naming: "get" is ambiguous, naively you'd read that as "just give me the current value" and not "increment and return the new value".  Just "`k_ref()`" as a function name would be very reasonable IMHO.  But "increment" and "reference" are probably better verbs than "get".

Reply:
Initially I wanted to use the usual ref/unref. But then I ended up with "k_ref_ref" and "k_ref_unref" which seemed kinda redundant. If sys_ref an and "sys_unref" are okay that'd work though.

Reply:
I renamed them to sys_ref and sys_unref